### ISSUE TITLE
How to work with WSL?

### ISSUE THREAD

**Issue by jtpavlock** [abo-abo/org-download#178]
Currently, it doesn't seem like there's anyway to utilize this package if using emacs in WSL on Windows. I was wondering if anyone had any success or tips for doing so?

**Comment by LesleyLai** [comment]
Yeah, this is such a nice package and I hope there is a way it can work with WSL GUI Emacs. I can still download from the URL, but there is no way to drag-and-drop local images.

**Comment by wang1zhen** [comment]
Also, the package is brilliant, but the `org-download-clipboard` is not working in wsl. Actually, it worked in Windows 10 with system screenshot tool, but after I upgraded to windows 11, `org-download-clipboard` generates a 0KB blank file from my system screenshot tool.

**Comment by liangzid** [comment]
I have the same problem, I use win10 + wsl2 + mobaxterm, and there is no way to drag or clip images.

**Comment by liangzid** [comment]
And I have find a solution with the help of windows Powershell, this is the solution function for replacing `org-download-clipboard`:
```emacs-lisp
(defun my-yank-image-from-win-clipboard-through-powershell()
  "to simplify the logic, use c:/Users/Public as temporary directoy, and move it into current directoy"
  (interactive)
  (let* ((powershell "/mnt/c/Windows/System32/WindowsPowerShell/v1.0/powershell.exe")
         (file-name (format-time-string "screenshot_%Y%m%d_%H%M%S.png"))
         ;; (file-path-powershell (concat "c:/Users/\$env:USERNAME/" file-name))
         (file-path-wsl (concat "./images/" file-name))
         )
    ;; (shell-command (concat powershell " -command \"(Get-Clipboard -Format Image).Save(\\\"C:/Users/\\$env:USERNAME/" file-name "\\\")\""))
    (shell-command (concat powershell " -command \"(Get-Clipboard -Format Image).Save(\\\"C:/Users/Public/" file-name "\\\")\""))
    (rename-file (concat "/mnt/c/Users/Public/" file-name) file-path-wsl)
    (insert (concat "[[file:" file-path-wsl "]]"))
    (message "insert DONE.")
    ))

```

**Comment by Awannaphasch2016** [comment]
> And I have find a solution with the help of windows Powershell, this is the solution function for replacing `org-download-clipboard`:
> 
> ```emacs-lisp
> (defun my-yank-image-from-win-clipboard-through-powershell()
>   "to simplify the logic, use c:/Users/Public as temporary directoy, and move it into current directoy"
>   (interactive)
>   (let* ((powershell "/mnt/c/Windows/System32/WindowsPowerShell/v1.0/powershell.exe")
>          (file-name (format-time-string "screenshot_%Y%m%d_%H%M%S.png"))
>          ;; (file-path-powershell (concat "c:/Users/\$env:USERNAME/" file-name))
>          (file-path-wsl (concat "./images/" file-name))
>          )
>     ;; (shell-command (concat powershell " -command \"(Get-Clipboard -Format Image).Save(\\\"C:/Users/\\$env:USERNAME/" file-name "\\\")\""))
>     (shell-command (concat powershell " -command \"(Get-Clipboard -Format Image).Save(\\\"C:/Users/Public/" file-name "\\\")\""))
>     (rename-file (concat "/mnt/c/Users/Public/" file-name) file-path-wsl)
>     (insert (concat "[[file:" file-path-wsl "]]"))
>     (message "insert DONE.")
>     ))
> ```

I have 0 knowledge about powershell, but from reading and experimenting, I found that the powershell command that you use doesn't save image and it also doesn't produce any error.  I use 5.1 version. Can you help me formulate powershell command that would work on 5.1 version?

**Comment by sdoyle88** [comment]
> 

I stumbled upon this as I was looking for a similar solution, and this function will work on WSL2 (I am running Ubuntu 20.04).  Ensure that the pathing is right, particularly to the powershell command (_/mnt/c/Windows/System32/..._), the _file-path-wsl_ _concat_ path, and the _shell_ and _rename_ commands.

For example, I changed my file-path-wsl to be "~/GDrive/emacs/org-download-images/", and changed that in the shell-command line too.

The powershell .exe hosted in System32 under v1.0 is Powershell 5.1 on Windows systems.  Powershell 6+ is Powershell core and usually stored under Program Files.  Also, the **Get-Clipboard** cmdlet in Powershell6+ doesnt support the Image format (yet), just Raw data, so you would want to utilize powershell 5.1 (as mentioned in the code above) if you want a built in cmdlet that can do that conversion.

**Comment by fkgruber** [comment]
I wanted to continue using org-download so I just changed the org-download-screenshot-method:

(use-package org-download
  :after org
  :bind
  (:map org-mode-map
        (("C-M-S-Y" . org-download-screenshot)
         ("s-y" . org-download-yank)))
  :custom
  (org-download-screenshot-method "powershell.exe -Command \"(Get-Clipboard -Format image).Save('$(wslpath -w %s)')\"")
  )

**Comment by fkgruber** [comment]
any way to be able to use drag and drop?
